### ISSUE TITLE
Wrong product url from getProductUrl when current category has not product object

### ISSUE THREAD

**Issue by OleksiiBulba** [magento/magento2#17819]
### Preconditions
1. Install Magento **2.2.4**
2. Create two category **category1** and **category2** with url-keys _category-1_ and _category-2_ respectively
3. Create product with url key _product-1_, assign it to **category1**
4. Add created categories to topmenu (turn on "include to menu" checkbox)

### Steps to reproduce
1. Create custom module Vendor_Module; run **setup:upgrade** command
2. Create template; add it to category layout
3. In the template get product collection with all products in it
4. In foreach loop try to get product url:
```
<?php foreach ($productCollection as $product): ?>
<? /** @var $product \Magento\Catalog\Model\Product */ ?>
<a href="<?= $product->getProductUrl() ?>"><?= $product->getName() ?></a>
<?php endforeach; ?>
```
5. Open **category2**(!) page  _http://example.com/category-2_

### Expected result
1. The product link should be like _http://example.com/category-1/product-1_

### Actual result
1. The product link is _http://example.com/catalog/product/view/id/1/s/product-1/category/2_

### Possible reason
1. In the file **app/code/Magento/Catalog/Model/Product.php:695** (2.2 branch) product tries to get category from current_category;
2. In the file **app/code/Magento/Catalog/Model/Product/Url.php:163** (2.2 branch) Url model is searching for url rewrite for **product1** from **category2**, but such rewrite does not exist because **product1** does not belong to **category2**

### Possible solution
1. Check if retrieved category has product
2. Add param to getProductUrl that specifies desired category in product url, for example:
```
<?php
$product->getProductUrl($category);
// or
$urlModel->getProductUrl($category, $product);
```

**Comment by magento-engcom-team** [comment]
Hi @OleksiiBulba. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me {$VERSION} instance` 

where `{$VERSION}` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (2.2-develop +). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@OleksiiBulba do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by engcom-backlog-nickolas** [comment]
Hello @OleksiiBulba, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by magento-engcom-team** [comment]
Hi @OleksiiBulba. Thank you for your report. 
The issue has been fixed in magento/magento2#19232 by @ErikPel in 2.3-develop branch
Related commit(s):
 - [a5dd6dc4c34d25a468b51ee383049a5ef0fc44f6](https://github.com/magento/magento2/commit/a5dd6dc4c34d25a468b51ee383049a5ef0fc44f6)

The fix will be available with the upcoming 2.3.1 release.

**Comment by magento-engcom-team** [comment]
Hi @OleksiiBulba. Thank you for your report. 
The issue has been fixed in magento/magento2#20286 by @GovindaSharma in 2.2-develop branch
Related commit(s):
 - [0247fc371ac7461cd35b8571dbedaef5a0a30834](https://github.com/magento/magento2/commit/0247fc371ac7461cd35b8571dbedaef5a0a30834)

The fix will be available with the upcoming 2.2.8 release.
